charmcraft/commands/store/__init__.py

```diff
@@ -461,3 +467,109 @@ def run(self, parsed_args):
+
+
+def _get_lib_info(*, full_name=None, lib_path=None):
+    """Get the whole lib info from the path/file."""
+    if full_name is None:
+        # get it from the lib_path
+        bad_structure_msg = (
+            "Library path {} must conform to the lib/charms/<charm>/v<API>/<libname>.py "
+            "structure.".format(lib_path))
+        try:
+            libsdir, charmsdir, charm_name, v_api = lib_path.parts[:-1]
+        except ValueError:
+            raise CommandError(bad_structure_msg)
+        if libsdir != 'lib' or charmsdir != 'charms' or lib_path.suffix != '.py':
+            raise CommandError(bad_structure_msg)
+        full_name = '.'.join((charmsdir, charm_name, v_api, lib_path.stem))
+
+    else:
+        # build the path! convert a lib name with dots to the full path, including lib
+        # dir and Python extension.
+        #    e.g.: charms.mycharm.v4.foo -> lib/charms/mycharm/v4/foo.py
+        bad_structure_msg = (
+            "Library full name {!r} must conform to the charms.<charm>.v<API>.<libname> structure."
+            .format(full_name))
+        try:
+            charmsdir, charm_name, v_api, libfile = full_name.split('.')
+        except ValueError:
+            raise CommandError(bad_structure_msg)
+        if charmsdir != 'charms':
+            raise CommandError(bad_structure_msg)
+        lib_path = pathlib.Path('lib') / charmsdir / charm_name / v_api / (libfile + '.py')
+
+    if v_api[0] != 'v' or not v_api[1:].isdigit():
+        raise CommandError(
+            "The API version in the library path must be 'vN' where N is an integer.")
+    api_from_path = int(v_api[1:])
+
+    lib_name = lib_path.stem
+    if not lib_path.exists():
+        return LibData(
+            lib_id=None, api=api_from_path, patch=-1, content_hash=None, content=None,
+            full_name=full_name, path=lib_path, lib_name=lib_name, charm_name=charm_name)
+
+    # parse the file and extract metadata from it, while hashing
+    metadata_fields = (b'LIBAPI', b'LIBPATCH', b'LIBID')
+    metadata = dict.fromkeys(metadata_fields)
+    hasher = hashlib.sha256()
+    with lib_path.open('rb') as fh:
+        for line in fh:
+            if line.startswith(metadata_fields):
+                try:
+                    field, value = [x.strip() for x in line.split(b'=')]
+                except ValueError:
+                    raise CommandError("Bad metadata line in {}: {!r}".format(lib_path, line))
+                metadata[field] = value
+            else:
+                hasher.update(line)
+
+    missing = [k.decode('ascii') for k, v in metadata.items() if v is None]
+    if missing:
+        raise CommandError(
+            "Library {} is missing the mandatory metadata fields: {}."
+            .format(lib_path, ', '.join(sorted(missing))))
+
+    def _get_positive_int(key):
+        """Convert the raw value for api/patch into a positive integer."""
+        value = metadata[key].decode('ascii')
```

Comment:
why the decode?
and why define this function on every call to _get_lib_info?

Reply:
Right, no need to decode!

And moved the function outside.